File: tests/test_timer_engine.py
"""Unit tests for JPred timer engine."""
import pytest
import time
import json
from pathlib import Path
from unittest.mock import patch

from jpred.timer_engine import TimerEngine, TimerType, Session


class TestTimerEngine:
    """Tests for TimerEngine class."""
    
    @pytest.fixture
    def engine(self, tmp_path):
        """Create a fresh engine with temp storage."""
        data_path = tmp_path / "test_data.json"
        return TimerEngine(data_path=data_path)
    
    def test_initial_state(self, engine):
        """Test initial timer states."""
        for timer_type in TimerType:
            assert engine.get_total_time(timer_type) == 0.0
            assert not engine.is_running(timer_type)
    
    def test_start_timer(self, engine):
        """Test starting a timer."""
        engine.start_timer(TimerType.JOB_SEARCH)
        assert engine.is_running(TimerType.JOB_SEARCH)
    
    def test_stop_timer(self, engine):
        """Test stopping a timer."""
        engine.start_timer(TimerType.PRACTICE)
        time.sleep(0.1)
        session = engine.stop_timer(TimerType.PRACTICE)
        
        assert not engine.is_running(TimerType.PRACTICE)
        assert session is not None
        assert session.duration >= 0.1
        assert engine.get_total_time(TimerType.PRACTICE) >= 0.1
    
    def test_reset_timer(self, engine):
        """Test resetting a timer."""
        engine.start_timer(TimerType.UPSKILLING)
        time.sleep(0.1)
        engine.stop_timer(TimerType.UPSKILLING)
        
        initial_time = engine.get_total_time(TimerType.UPSKILLING)
        assert initial_time > 0
        
        engine.reset_timer(TimerType.UPSKILLING)
        assert engine.get_total_time(TimerType.UPSKILLING) == 0.0
    
    def test_elapsed_time_while_running(self, engine):
        """Test elapsed time calculation for running timer."""
        engine.start_timer(TimerType.JOB_SEARCH)
        time.sleep(0.2)
        
        elapsed = engine.get_elapsed_time(TimerType.JOB_SEARCH)
        assert elapsed >= 0.2
    
    def test_persistence(self, engine, tmp_path):
        """Test data persistence."""
        data_path = tmp_path / "test_data.json"
        
        # Start and stop a timer
        engine.start_timer(TimerType.JOB_SEARCH)
        time.sleep(0.1)
        engine.stop_timer(TimerType.JOB_SEARCH)
        
        # Create new engine with same path
        new_engine = TimerEngine(data_path=data_path)
        
        # Verify data persisted
        assert new_engine.get_total_time(TimerType.JOB_SEARCH) >= 0.1
        assert len(new_engine.get_all_sessions()[TimerType.JOB_SEARCH]) == 1
    
    def test_goals(self, engine):
        """Test goal setting and retrieval."""
        engine.set_goal(TimerType.JOB_SEARCH, 60)
        assert engine.get_goal(TimerType.JOB_SEARCH) == 60
        
        engine.set_goal(TimerType.PRACTICE, 90)
        assert engine.get_goal(TimerType.PRACTICE) == 90
    
    def test_progress_calculation(self, engine):
        """Test progress towards goals."""
        engine.set_goal(TimerType.JOB_SEARCH, 60)  # 60 minutes goal
        
        # 0% progress initially
        assert engine.get_progress(TimerType.JOB_SEARCH) == 0.0
        
        # Add 30 minutes (50% progress)
        with patch('time.time') as mock_time:
            mock_time.return_value = 1000
            engine.start_timer(TimerType.JOB_SEARCH)
            mock_time.return_value = 1000 + 1800  # 30 minutes later
            engine.stop_timer(TimerType.JOB_SEARCH)
        
        assert engine.get_progress(TimerType.JOB_SEARCH) == 0.5
    
    def test_session_history(self, engine):
        """Test session history tracking."""
        # Create multiple sessions
        for _ in range(3):
            engine.start_timer(TimerType.PRACTICE)
            time.sleep(0.05)
            engine.stop_timer(TimerType.PRACTICE)
        
        sessions = engine.get_all_sessions()[TimerType.PRACTICE]
        assert len(sessions) == 3


class TestSession:
    """Tests for Session dataclass."""
    
    def test_session_creation(self):
        """Test creating a session."""
        session = Session(
            start_time=1000.0,
            end_time=1100.0,
            duration=100.0
        )
        
        assert session.start_time == 1000.0
        assert session.end_time == 1100.0
        assert session.duration == 100.0
    
    def test_session_serialization(self):
        """Test session to_dict and from_dict."""
        original = Session(
            start_time=1000.0,
            end_time=1100.0,
            duration=100.0
        )
        
        data = original.to_dict()
        restored = Session.from_dict(data)
        
        assert restored.start_time == original.start_time
        assert restored.end_time == original.end_time
        assert restored.duration == original.duration


class TestTimerType:
    """Tests for TimerType enum."""
    
    def test_timer_types(self):
        """Test all timer types exist."""
        assert TimerType.JOB_SEARCH.value == "job_search"
        assert TimerType.PRACTICE.value == "practice"
        assert TimerType.UPSKILLING.value == "upskilling"


if __name__ == "__main__":
    pytest.main([__file__, "-v"])

File: tests/test_timer_engine.js
const assert = require('assert');
const { TimerEngine, TimerType } = require('../src/timerEngine.js');
const fs = require('fs');
const path = require('path');

// Test helper - create temp data path
function createTempPath() {
  return path.join(process.env.TEMP || '/tmp', 'jpred_test_' + Date.now() + '.json');
}

// Clean up temp file
function cleanupTemp(path) {
  try {
    if (fs.existsSync(path)) {
      fs.unlinkSync(path);
    }
  } catch (e) {}
}

console.log('Running JPred Timer Engine Tests...\n');

let passed = 0;
let failed = 0;

function test(name, fn) {
  try {
    fn();
    console.log('✓ ' + name);
    passed++;
  } catch (error) {
    console.log('✗ ' + name);
    console.log('  Error: ' + error.message);
    failed++;
  }
}

// Test 1: Initial state
test('Initial timer states are zero', () => {
  const tempPath = createTempPath();
  const engine = new TimerEngine(tempPath);
  
  assert.strictEqual(engine.getElapsedTime(TimerType.JOB_SEARCH), 0);
  assert.strictEqual(engine.getElapsedTime(TimerType.PRACTICE), 0);
  assert.strictEqual(engine.getElapsedTime(TimerType.UPSKILLING), 0);
  
  cleanupTemp(tempPath);
});

// Test 2: Start timer
test('Start timer sets isRunning to true', () => {
  const tempPath = createTempPath();
  const engine = new TimerEngine(tempPath);
  
  engine.startTimer(TimerType.JOB_SEARCH);
  assert.strictEqual(engine.isRunning(TimerType.JOB_SEARCH), true);
  
  cleanupTemp(tempPath);
});

// Test 3: Stop timer
test('Stop timer creates session and adds time', (done) => {
  const tempPath = createTempPath();
  const engine = new TimerEngine(tempPath);
  
  engine.startTimer(TimerType.PRACTICE);
  
  setTimeout(() => {
    const session = engine.stopTimer(TimerType.PRACTICE);
    
    assert.strictEqual(engine.isRunning(TimerType.PRACTICE), false);
    assert(session !== null);
    assert(session.duration >= 0.1);
    assert(engine.getElapsedTime(TimerType.PRACTICE) >= 0.1);
    
    cleanupTemp(tempPath);
  }, 150);
});

// Test 4: Reset timer
test('Reset timer clears total time', () => {
  const tempPath = createTempPath();
  const engine = new TimerEngine(tempPath);
  
  // Manually set some time
  engine.timers[TimerType.UPSKILLING].totalTime = 100;
  engine.save();
  
  engine.resetTimer(TimerType.UPSKILLING);
  assert.strictEqual(engine.getElapsedTime(TimerType.UPSKILLING), 0);
  
  cleanupTemp(tempPath);
});

// Test 5: Goals
test('Set and get goals', () => {
  const tempPath = createTempPath();
  const engine = new TimerEngine(tempPath);
  
  engine.setGoal(TimerType.JOB_SEARCH, 60);
  assert.strictEqual(engine.getGoal(TimerType.JOB_SEARCH), 60);
  
  engine.setGoal(TimerType.PRACTICE, 90);
  assert.strictEqual(engine.getGoal(TimerType.PRACTICE), 90);
  
  cleanupTemp(tempPath);
});

// Test 6: Progress calculation
test('Progress calculation', () => {
  const tempPath = createTempPath();
  const engine = new TimerEngine(tempPath);
  
  engine.setGoal(TimerType.JOB_SEARCH, 60); // 60 minutes goal
  
  // Initially 0% progress
  assert.strictEqual(engine.getProgress(TimerType.JOB_SEARCH), 0);
  
  // Manually set 30 minutes (50% progress)
  engine.timers[TimerType.JOB_SEARCH].totalTime = 30 * 60;
  assert.strictEqual(engine.getProgress(TimerType.JOB_SEARCH), 0.5);
  
  cleanupTemp(tempPath);
});

// Test 7: Session history
test('Session history tracking', () => {
  const tempPath = createTempPath();
  const engine = new TimerEngine(tempPath);
  
  // Create multiple sessions
  for (let i = 0; i < 3; i++) {
    engine.startTimer(TimerType.PRACTICE);
    engine.stopTimer(TimerType.PRACTICE);
  }
  
  const sessions = engine.getAllSessions()[TimerType.PRACTICE];
  assert.strictEqual(sessions.length, 3);
  
  cleanupTemp(tempPath);
});

// Test 8: Persistence
test('Data persistence', () => {
  const tempPath = createTempPath();
  
  // Create engine and add data
  const engine1 = new TimerEngine(tempPath);
  engine1.startTimer(TimerType.JOB_SEARCH);
  engine1.stopTimer(TimerType.JOB_SEARCH);
  engine1.setGoal(TimerType.JOB_SEARCH, 45);
  
  // Create new engine with same path
  const engine2 = new TimerEngine(tempPath);
  
  // Verify data persisted
  assert(engine2.getElapsedTime(TimerType.JOB_SEARCH) >= 0);
  assert.strictEqual(engine2.getGoal(TimerType.JOB_SEARCH), 45);
  
  cleanupTemp(tempPath);
});

// Test 9: Today's sessions
test('Get today\'s sessions', () => {
  const tempPath = createTempPath();
  const engine = new TimerEngine(tempPath);
  
  engine.startTimer(TimerType.JOB_SEARCH);
  engine.stopTimer(TimerType.JOB_SEARCH);
  
  const todaySessions = engine.getTodaySessions(TimerType.JOB_SEARCH);
  assert(todaySessions.length >= 1);
  
  cleanupTemp(tempPath);
});

// Test 10: Format time helper
test('Format time helper', () => {
  // Test MM:SS format
  let result = formatTime(125); // 2 minutes 5 seconds
  assert.strictEqual(result, '02:05');
  
  // Test HH:MM:SS format
  result = formatTime(3661); // 1 hour 1 minute 1 second
  assert.strictEqual(result, '01:01:01');
});

function formatTime(seconds) {
  const hours = Math.floor(seconds / 3600);
  const minutes = Math.floor((seconds % 3600) / 60);
  const secs = Math.floor(seconds % 60);
  
  if (hours > 0) {
    return String(hours).padStart(2, '0') + ':' + String(minutes).padStart(2, '0') + ':' + String(secs).padStart(2, '0');
  }
  return String(minutes).padStart(2, '0') + ':' + String(secs).padStart(2, '0');
}

// Summary
console.log('\n' + '='.repeat(40));
console.log('Tests: ' + (passed + failed) + ' | Passed: ' + passed + ' | Failed: ' + failed);
console.log('='.repeat(40));

process.exit(failed > 0 ? 1 : 0);
